Introduce GGML_CALL function annotation

jart opened the pull request:
This change makes it possible to build ggml-cuda.cu and ggml-metal.m as independent dynamic shared objects, that may be conditionally linked at runtime in a multiplatform binary. It introduces a GGML_CALL annotation that documents which functions have a cyclic call relationship, between the application code and GPU modules.

This change does nothing, unless the build defines -DGGML_MULTIPLATFORM which causes back-references and function pointers to conform to MS ABI which is supported by NVCC, ROCm, XCode, GCC and Clang across platforms.

Additional context:

- https://github.com/Mozilla-Ocho/llamafile/commit/20d5f465285332dade7e445c44f7559953b8ad39
- https://github.com/Mozilla-Ocho/llamafile/commit/4616816aacba97ff3849e25f9f0fc814f21b9c7a

ggerganov commented:
Need to resolve the conflicts due to the recent API changes

cebtenzzre commented:
This seems like a pretty esoteric use case. llamafile would be the only downstream user of this.

jart committed:
Introduce GGML_CALL function annotation

This change makes it possible to build ggml-cuda.cu and ggml-metal.m as
independent dynamic shared objects, that may be conditionally linked at
runtime in a multiplatform binary. It introduces a GGML_CALL annotation
that documents which functions have a cyclic call relationship, between
the application code and GPU modules.

This change does nothing, unless the build defines -DGGML_MULTIPLATFORM
which causes back-references and function pointers to conform to MS ABI
which is supported by NVCC, ROCm, XCode, GCC and Clang across platforms

jart commented:
@ggerganov Done. I've merged 1633 lines of merge conflicts. PTAL

As for @cebtenzzre concerns, llamafile users are llama.cpp users. Please consider that we don't white label the llama.cpp software, since we use the upstream branding in the web interface. Our project is simply focused on binary distributability. By merging this change, you'll be helping us to deliver the most up-to-date and accurate representation of the llama.cpp software possible.

ggerganov commented:
I've been thinking for a while and I agree that it's a quite limited use case for the amount of changes to the library. But I do find the `llamafile` project to be a very cool technical feat, so I'm okay to incorporate `GGML_CALL` in order to support it. Hopefully it will not interfere with our development - it looks like a pretty transparent change.

I'm not familiar how the `__attribute__((__ms_abi__))` works, but in case it could be prefixed instead of suffixed it would be better to move it next to `GGML_API` in the headers.

Unless there are some other considerations to be discussed, we can proceed with merging

jart commented:
@ggerganov It can be adjacent to `GGML_API`. Would you like me to adjust the indentation too?

ggerganov commented:
Yes, adjust it just within blocks of declarations that have `GGML_API GGML_CALL`

```c
// example
GGML_API GGML_CALL void ...
GGML_API           void ...

GGML_API void ...
```

jart commented:
@ggerganov Updated. Tested with define enabled for both Linux Cuda and Apple Metal. PTAL.
